galley/pkg/config/analysis/analyzers/auth/jwt.go

```diff
@@ -0,0 +1,170 @@
+	return analysis.Metadata{
+		Name:        "injection.JwtAnalyzer",
+		Description: "Checks that jwt auth policies are against valid services",
+		Inputs: collection.Names{
+			collections.IstioAuthenticationV1Alpha1Policies.Name(),
+			collections.K8SCoreV1Services.Name(),
+		},
+	}
+}
+
+func (j *JwtAnalyzer) Analyze(c analysis.Context) {
+	nsm := j.buildNamespaceServiceMap(c)
+
+	c.ForEach(collections.IstioAuthenticationV1Alpha1Policies.Name(), func(r *resource.Instance) bool {
+		j.analyzeServiceTarget(r, c, nsm)
+		return true
+	})
+}
+
+// buildNamespaceServiceMap returns a map where the index is a namespace and the boolean
+func (j *JwtAnalyzer) buildNamespaceServiceMap(ctx analysis.Context) map[string]*v1.ServiceSpec {
+	// Keep track of each fqdn -> service definition
+	fqdnServices := map[string]*v1.ServiceSpec{}
+
+	ctx.ForEach(collections.K8SCoreV1Services.Name(), func(r *resource.Instance) bool {
+		svcNs := r.Metadata.FullName.Namespace
+		svcName := r.Metadata.FullName.Name
+
+		svc := r.Message.(*v1.ServiceSpec)
+		fqdn := util.ConvertHostToFQDN(svcNs, string(svcName))
+		fqdnServices[fqdn] = svc
+
+		return true
+	})
+
+	return fqdnServices
+}
+
+func (j *JwtAnalyzer) analyzeServiceTarget(r *resource.Instance, ctx analysis.Context, nsm map[string]*v1.ServiceSpec) {
+	policy := r.Message.(*v1alpha1.Policy)
+	ns := r.Metadata.FullName.Namespace
+
+	for _, origin := range policy.Origins {
+		if origin.GetJwt() != nil {
+			for _, target := range policy.GetTargets() {
+
+				fqdn := util.ConvertHostToFQDN(ns, target.GetName())
+				if svc, has := nsm[fqdn]; has {
+					if len(target.GetPorts()) == 0 {
+						checkServicePorts(r, ctx, svc)
+						continue
+					}
+
+					// check ports defined in the authentication policy for the service
+					for _, port := range target.GetPorts() {
+						if port.GetName() == "" {
+							checkPort(r, ctx, port.GetNumber(), svc)
+						}
+
+					}
+				} // else no target service was found, but not an error
+			}
+		}
+	}
+
+}
+
+func checkPort(r *resource.Instance, ctx analysis.Context, portNum uint32, svc *v1.ServiceSpec) {
+	var svcPort *v1.ServicePort
+	for _, port := range svc.Ports {
+		if strings.ToUpper(string(port.Protocol)) != "TCP" && port.Protocol != "" {
+			continue
+		}
+		if portNum == uint32(port.Port) {
+			svcPort = &port
+			break
+		}
+	}
+	if svcPort != nil {
+		portName := strings.ToLower(svcPort.Name)
+		if !portHasSupportedPrefix(portName) {
+			ctx.Report(collections.IstioAuthenticationV1Alpha1Policies.Name(),
+				msg.NewJwtFailureDueToInvalidServicePortPrefix(
+					r,
+					int(portNum),
+					portName,
+					string(svcPort.Protocol),
+					svcPort.TargetPort.String(),
+				))
+		}
+
+	}
+}
+
+func checkServicePorts(r *resource.Instance, ctx analysis.Context, svc *v1.ServiceSpec) {
+	for _, port := range svc.Ports {
+		portName := strings.ToLower(port.Name)
+		if (strings.ToUpper(string(port.Protocol)) == "TCP" || port.Protocol == "") &&
```

Comment:
Can you normalize to upper or lower case consistently in this function?

Reply:
Added a predicate function to check if the port is TCP or not.

I was able to normalize as K8s enforces that the protocol must be upper case and enforces that names are lower case.